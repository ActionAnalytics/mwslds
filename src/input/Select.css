
.Select-control {
  border-radius: 0;
}

/* nasty */
.Select.has-value.Select--single > .Select-control .Select-value .Select-value-label, .Select.has-value.is-pseudo-focused.Select--single > .Select-control .Select-value .Select-value-label {
  color: #555;
}
