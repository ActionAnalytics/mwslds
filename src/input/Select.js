import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { startCase } from 'lodash'

import '../style'

const propTypes = {
  name: PropTypes.string.isRequired,
  data: PropTypes.arrayOf(PropTypes.object),
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  prefix: PropTypes.string,
  disabled: PropTypes.bool,
  multi: PropTypes.bool,
  width: PropTypes.string,
  validationError: PropTypes.string,
}

const defaultProps = {
  data: [],
  prefix: null,
  disabled: null,
  multi: null,
  width: null,
  validationError: null,
}

function onChangeWrapper(wrapped) {
  return evt => wrapped(evt.target.value)
}

function Select(props) {
  const {
    name,
    data,
    value,
    onChange,
    prefix,
    disabled,
    multi,
    width,
    validationError,
  } = props

  const title = startCase(name)
  const inputId = `${prefix || ''}${name}`
  const labelClasses = classnames('control-label', validationError && 'has-error')

  return (
    <label className={labelClasses} htmlFor={inputId} style={{ width }}>
      {title}
      <select
        className="form-control"
        value={value}
        onChange={onChangeWrapper(onChange)}
        disabled={disabled}
        multiple={multi}
      >
        {
          data.map(item => (
            item && <option key={item.value} value={item.value}>{item.name}</option>
          ))
        }
      </select>
    </label>
  )
}

Select.propTypes = propTypes
Select.defaultProps = defaultProps

export default Select
