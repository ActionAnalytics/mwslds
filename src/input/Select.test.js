import React from 'react'
import { configure, shallow, mount } from 'enzyme'
import { createSerializer } from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'

import Select from './Select'

configure({ adapter: new Adapter() })
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

describe('Select', () => {
  it('renders', () => {
    const select = mount(<Select name="selectceptional" onChange={() => {}} />)
    expect(select).toMatchSnapshot()
  })

  it('has the correct title', () => {
    const select = shallow(<Select name="selectceptional" title="not selectceptional" onChange={() => {}} />)
    expect(select).toMatchSnapshot()
  })

  it('has the correct width', () => {
    const wide = shallow(<Select name="selectceptional" width="100%" onChange={() => {}} />)
    expect(wide).toMatchSnapshot()

    const lesswide = shallow(<Select name="selectceptional" width="50%" onChange={() => {}} />)
    expect(lesswide).toMatchSnapshot()
  })

  it('renders options', () => {
    const opts = [
      { value: 'a', label: 'A' },
      { value: 'b', label: 'B' },
      { value: 'c', label: 'C' },
    ]
    const select = mount(<Select name="selectceptional" data={opts} onChange={() => {}} />)
    expect(select).toMatchSnapshot()
  })

  it('displays the correct values', () => {
    const opts = [
      { value: 'a', label: 'A' },
      { value: 'b', label: 'B' },
      { value: 'c', label: 'C' },
    ]
    const selected = 'a'
    const select = mount(<Select name="selectceptional" data={opts} value={selected} onChange={() => {}} />)
    expect(select).toMatchSnapshot()
    expect(select.find('VirtualizedSelect').props().value).toBe(selected)
  })

  it('returns value for change events', () => {
    const select = shallow(<Select name="selectceptional" value="whatever" onChange={val => val} />)
    const input = select.find('VirtualizedSelect')
    expect(input.props().onChange({ value: 'hi' })).toBe('hi')
    expect(input.props().onChange(null)).toBe('')
  })

  it('is disabled properly', () => {
    const disabled = shallow(<Select name="selectceptional" disabled onChange={() => {}} />)
    expect(disabled.find('VirtualizedSelect').props().disabled).toBe(true)

    const readOnly = shallow(<Select name="selectceptional" readOnly onChange={() => {}} />)
    expect(readOnly.find('VirtualizedSelect').props().disabled).toBe(true)
  })

  it('shows errors properly', () => {
    const error = mount(<Select name="selectceptional" validationError="whatever" onChange={() => {}} />)
    expect(error).toMatchSnapshot()
  })
})
