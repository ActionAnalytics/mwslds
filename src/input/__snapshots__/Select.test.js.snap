// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select displays the correct values 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": null,
    }
  }
>
  Selectceptional
  <div
    className="Select has-value is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-4--value"
      >
        <div
          className="Select-value"
        >
          <span
            aria-selected="true"
            className="Select-value-label"
            id="react-select-4--value-item"
            role="option"
          >
            A
          </span>
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-4--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            id="selectceptional"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        aria-label="Clear value"
        className="Select-clear-zone"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        title="Clear value"
      >
        <span
          className="Select-clear"
          dangerouslySetInnerHTML={
            Object {
              "__html": "&times;",
            }
          }
        />
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`Select has the correct title 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": null,
    }
  }
>
  not selectceptional
  <VirtualizedSelect
    async={false}
    className={null}
    disabled={null}
    id="selectceptional"
    maxHeight={200}
    multiple={null}
    onChange={[Function]}
    optionHeight={35}
    options={Array []}
    selectComponent={[Function]}
  />
</label>
`;

exports[`Select has the correct width 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": "100%",
    }
  }
>
  Selectceptional
  <VirtualizedSelect
    async={false}
    className={null}
    disabled={null}
    id="selectceptional"
    maxHeight={200}
    multiple={null}
    onChange={[Function]}
    optionHeight={35}
    options={Array []}
    selectComponent={[Function]}
  />
</label>
`;

exports[`Select has the correct width 2`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": "50%",
    }
  }
>
  Selectceptional
  <VirtualizedSelect
    async={false}
    className={null}
    disabled={null}
    id="selectceptional"
    maxHeight={200}
    multiple={null}
    onChange={[Function]}
    optionHeight={35}
    options={Array []}
    selectComponent={[Function]}
  />
</label>
`;

exports[`Select renders 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": null,
    }
  }
>
  Selectceptional
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-2--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-2--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            id="selectceptional"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`Select renders options 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": null,
    }
  }
>
  Selectceptional
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-3--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-3--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            id="selectceptional"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`Select shows errors properly 1`] = `
<label
  className="control-label"
  htmlFor="selectceptional"
  style={
    Object {
      "width": null,
    }
  }
>
  Selectceptional
  <div
    className="Select has-error is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-5--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-5--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            id="selectceptional"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;
